### PR TITLE
Bug report server script

### PR DESCRIPTION
**Pull request by Kegsay**
Example logs (`-0` is most recent, `-3` is oldest):

```
$ ls -l
-rw-r--r--  1 kegan  staff   909 26 Jan 11:46 bugreport-20170126-114615-0.log.gz
-rw-r--r--  1 kegan  staff   524 26 Jan 11:46 bugreport-20170126-114615-1.log.gz
-rw-r--r--  1 kegan  staff  1761 26 Jan 11:46 bugreport-20170126-114615-2.log.gz
-rw-r--r--  1 kegan  staff   998 26 Jan 11:46 bugreport-20170126-114615-3.log.gz
-rw-r--r--  1 kegan  staff   198 26 Jan 11:46 bugreport-20170126-114615.log.gz

$ zless bugreport-20170126-114615.log.gz 
everything is awful.

Number of logs: 4
Version: UNKNOWN
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.95 Safari/537.36

$ zless bugreport-20170126-114615-0.log.gz 
2017-01-26T10:52:37.374Z I Initialised rageshake: See https://bugs.chromium.org/p/chromium/issues/detail?id=583193 to fix line numbers on Chrome.
2017-01-26T10:52:37.376Z I Vector starting at http://localhost:8080/#/room/!OvylurBpcFVXBQyEGv:localhost
2017-01-26T10:52:37.400Z I Routing URL http://localhost:8080/#/room/!OvylurBpcFVXBQyEGv:localhost
2017-01-26T10:52:37.408Z I Restoring session for %s @foobar2:localhost
2017-01-26T10:52:37.408Z I setLoggedIn => %s (guest=%s) hs=%s @foobar2:localhost false http://localhost:8008
2017-01-26T10:52:37.409Z I Session persisted for %s @foobar2:localhost
2017-01-26T10:52:37.549Z I SyncApi.sync: starting with sync token null
2017-01-26T10:52:37.554Z E Failed to poll for update [object Object]
2017-01-26T10:52:37.560Z I newscreen 
2017-01-26T10:52:37.573Z I Presence: %s online
2017-01-26T10:52:37.574Z I Got push rules
2017-01-26T10:52:37.677Z I MatrixClient sync state => %s PREPARED
2017-01-26T10:52:37.681Z I MatrixClient sync state => %s SYNCING
2017-01-26T10:52:37.683Z I newscreen room/!OvylurBpcFVXBQyEGv:localhost
2017-01-26T10:52:37.943Z I updateTint from RoomView._gatherTimelinePanelRef
2017-01-26T10:52:37.945Z I Tinter.tint from updateTint
2017-01-26T10:52:40.650Z I newscreen settings
2017-01-26T10:52:40.821Z I Failed to fetch app version [object Object]
2017-01-26T10:52:45.674Z I Sending bug report.
2017-01-26T10:52:48.065Z E Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the RoomView component.
2017-01-26T10:56:30.634Z I Presence: %s unavailable
2017-01-26T11:02:38.610Z E Failed to poll for update [object Object]
2017-01-26T11:10:12.410Z I Presence: %s online
2017-01-26T11:10:15.215Z I Sending bug report.
...
```

### CODE REVIEW

**Comment by ara4n** [comment]
We need to make sure the reports go into a bug or ticket tracker. I suggest routing them into a private GH repo. Alternatively i have an almost complete install of RT on ldc-prd-matrix-001

-- 
Matthew Hodgson
Matrix.org

File: scripts/rageshake.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+			respond(400, w)
+			return
+		}
+		// Dump bug report to disk as form:
+		//  "bugreport-20170115-112233.log.gz" => user text, version, user agent, # logs
+		//  "bugreport-20170115-112233-0.log.gz" => most recent log
+		//  "bugreport-20170115-112233-1.log.gz" => ...
+		//  "bugreport-20170115-112233-N.log.gz" => oldest log
+		t := time.Now()
+		prefix := t.Format("bugreport-20060102-150405")
```
Comment (richvdh) [inline scripts/rageshake.go]: i assume this will be UTC?

**Comment by Kegsay** [comment]
What are the file size limits like when making issues on Github? We can obviously squeeze more out by gzipping like this does, but if it's an attachment then it's more faff to view (click to download, run through `gzip` or `zless` it rather than just clicking the attachment to view in the browser).

LMK where you want 'em and I'll send them in whatever form you want.

File: scripts/rageshake.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	Logs      []LogEntry `json:"logs"`
+}
+
+func respond(code int, w http.ResponseWriter) {
+	w.WriteHeader(code)
+	w.Write([]byte("{}"))
+}
+
+func gzipAndSave(data []byte, filepath string) error {
+	if _, err := os.Stat(filepath); err == nil {
```
Comment (richvdh) [inline scripts/rageshake.go]: should we be specifying a directory as an explicit option rather than just using the cwd?

**Comment by Kegsay** [comment]
Hmm Github are suitably vague, mainly focusing on the limits for git repos: https://help.github.com/articles/what-is-my-disk-quota/

**Comment by ara4n** [comment]
Suspect GH don't penalise size on private repos. I suggest we route them to GH for now, given it's not a very tickety thing - we're more likely to follow up by matrix than email!

-- 
Matthew Hodgson
Matrix.org

File: scripts/rageshake.go (hunk 1 above)
Comment (Kegsay) [inline scripts/rageshake.go]: It won't actually, will fix.

File: scripts/rageshake.go (hunk 2 above)
Comment (Kegsay) [inline scripts/rageshake.go]: ¯\\_(ツ)_/¯ maybe?

**Comment by Kegsay** [comment]
Oh, found https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/ - so 25MB and .gz WFM.

File: scripts/rageshake.go (hunk 1 above)
Comment (Kegsay) [inline scripts/rageshake.go]: Done.

**Comment by Kegsay** [comment]
@ara4n Github's API does not allow you to upload attachments to issues, so we cannot realistically send bug reports to GH issues.

**Comment by Kegsay** [comment]
Merging as per:

> kegan: wdyt about a cheap & cheerful soln to unblock rageshake? vdh's suggestion of just bunging them on disk for now

**Pull request merged**